### ISSUE TITLE
Never returns functions.

### ISSUE THREAD

**Issue by 0x0G** [Manishearth/rust-clippy#815]
If try to check this code:
```rust
let loaded = load().unwrap_or_else(|e| exit_with_error(e));

fn exit_with_error<E: std::error::Error>(e: E) -> ! {
    let _ = writeln!(&mut io::stderr(), "Error: {}", e);
    exit(-1)
}
```
Output warning:
```shell
warning: redundant closure found, #[warn(redundant_closure)] on by default
```
But if we change it to:
```rust
let loaded = load().unwrap_or_else(exit_with_error);
```
Will output error:
```
error: the trait `core::ops::FnOnce<(error::Error,)>` is not implemented for the type `fn(_) -> ! {exit_with_error::<_>}` [E0277]
```

**Comment by llogiq** [comment]
I think the problem lies with the generics, not the divergence of the function.

**Comment by mcarton** [comment]
@llogiq Nope. See #817, [the divergent function cannot be passed as `.map(divergent)`, but generics are not a problem](REDUNDANT_CLOSURE).

**Comment by llogiq** [comment]
I stand corrected. :smile:
